### DISCUSSION TITLE
console effect handler?

### DISCUSSION THREAD

**Discussion by woojamon** [koka-lang/koka#451] [General]
Koka doesn't seem to like me trying to handle the `console` effect, but perhaps I'm just not doing something correctly?

```
fun main() : _ ()

  var x := ""

  with handler 
    fun string/println(msg) x := msg

  println("hello")
```
> *** internal compiler error: Type.Infer.inferHandledEffect: cannot find handled effect in TApp (TCon (TypeCon {typeconName = std/core/types/effect-extend, typeconKind = KApp (KApp (KCon (->)) (KCon X)) (KApp (KApp (KCon (->)) (KCon E)) (KCon E))})) [TCon (TypeCon {typeconName = std/core/console, typeconKind = KCon X}),TVar (TypeVar {typevarId = 36, typevarKind = KCon E, typevarFlavour = Meta})]%           

And if I try:
```
fun main() : _ ()

  var x := ""

  with handler <console>
    fun string/println(msg) x := msg

  println("hello")
```
> identifier @Hnd-console cannot be found

**Comment by TimWhiting** [comment]
Console is specifically for something that really writes to the console (it's a builtin effect, that is not handleable).

What you probably want to do is create a `writer` effect, and then sometimes handle it yourself (for example unit tests), and sometimes actually write to the console, (insert a default handler that just forwards the writes to the actual console). Also quick tip: `trace` can be used for debugging functions without changing the effect type. (It just pretends that there is not an effect - since it is meant for debugging, though there really is).
